core/rawdb/accessors_chain.go

```diff
@@ -329,21 +329,44 @@ func ReadHeaderRange(db ethdb.Reader, number uint64, count uint64) []rlp.RawValu
 	return rlpHeaders
 }
 
+// HashByHeader is a compile-time constant, which can be used to toggle the behaviour of ReadHeaderRLP. If `false`,
+// it is assumed that the header hash == `crypto.Keccak256(header_rlp)`. This may not be true in all ethereum-forks.
+// Such forks need to set this constant to `true`, to force the hash-check to utilize the `header.Hash` method.
+const HashByHeader = false
+
 // ReadHeaderRLP retrieves a block header in its raw RLP database encoding.
 func ReadHeaderRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValue {
 	var data []byte
-	db.ReadAncients(func(reader ethdb.AncientReader) error {
+	err := db.ReadAncients(func(reader ethdb.AncientReader) error {
 		// First try to look up the data in ancient database. Extra hash
 		// comparison is necessary since ancient database only maintains
 		// the canonical data.
 		data, _ = reader.Ancient(freezerHeaderTable, number)
-		if len(data) > 0 && crypto.Keccak256Hash(data) == hash {
-			return nil
+		if len(data) > 0 {
+			if HashByHeader {
+				var header *types.Header
+				err := rlp.DecodeBytes(data, &header)
+				if err != nil {
+					return err
+				}
+				if header != nil && header.Hash() == hash {
+					return nil
+				}
+			} else {
+				if crypto.Keccak256Hash(data) == hash {
+					return nil
+				} else {
+					log.Warn("ancient header exists but rlpHash doesn't match the hash passed in, ignored, you may want to change HashByHeader to true")
```

Comment:
Nah, let's drop this warning. What would be nice, though, is to have a check in `init` (somewhere), that would panic if `HashByHeader` is false and `crypto.Keccak256(rlp(Header{})) != Header{}.Hash()` (pseudo-code)